crates/matrix-sdk/src/widget/machine/mod.rs

```diff
@@ -249,9 +240,35 @@ impl WidgetMachine {
         };
 
         match request {
-            ReadEventRequest::ReadMessageLikeEvent { .. } => {
-                let text = "Reading of message events is not yet supported";
-                self.send_from_widget_error_response(raw_request, text)
+            ReadEventRequest::ReadMessageLikeEvent { event_type, limit } => {
+                if !capabilities
+                    .read
+                    .iter()
+                    .any(|filter| filter.matches_message_like_event_type(&event_type))
+                {
+                    self.send_from_widget_error_response(raw_request, "Not allowed")
+                } else {
+                    const DEFAULT_EVENT_LIMIT: u32 = 50;
+                    let limit = limit.unwrap_or(DEFAULT_EVENT_LIMIT);
+                    let request = ReadMessageLikeEventRequest { event_type, limit };
+                    let (request, action) = self.send_matrix_driver_request(request);
+                    request.then(|result, machine| {
+                        let response = result.and_then(|events| {
+                            let CapabilitiesState::Negotiated(capabilities) = &machine.capabilities else {
+                                return Err("Received read event request before capabilities were negotiated".into());
+                            };
+
+                            Ok(ReadEventResponse {
+                                events: events
+                                    .into_iter()
+                                    .filter(|e| capabilities.raw_event_matches_read_filter(e))
+                                    .collect(),
```

Comment:
Instead of `into_iter` + `filter` + `collect`, you can use `.retain`.

Reply:
You're right! Fixed.